# Projeto Final

Grupo:

Bruno Santos - bfss@cesar.school

Gabriel Leite - gpl@cesar.school

Leonardo Santos - llss@cesar.school

Renata Pereira - rmbp@cesar.school

# 1. O que é Pokémon?

Os Pokémon são criaturas fictícias que habitam um mundo fictício chamado de Mundo Pokémon. Eles são os principais protagonistas da franquia de mídia Pokémon, que inclui jogos de vídeo, desenhos animados, filmes, card games e uma variedade de outros produtos.

Cada Pokémon é caracterizado por suas próprias individualidades, incluindo habilidades, status, egg groups e tipos:

- Habilidades: As habilidades dos Pokémon são características distintas que podem afetar o curso das batalhas ou interações no jogo. Algumas habilidades aumentam os atributos do Pokémon, outras alteram os efeitos dos movimentos ou têm efeitos passivos.
- Status: Os status dos Pokémon referem-se aos seus atributos básicos, como HP (pontos de vida), ataque, defesa, sp_attack (ataque especial), sp_defense (defesa especial) e velocidade. Cada Pokémon tem seus próprios valores de status, que podem variar amplamente entre as espécies.
- Egg Groups: Os egg groups definem com quais outros Pokémon um determinado Pokémon pode acasalar para produzir ovos. Esses grupos são usados para determinar a compatibilidade de reprodução entre Pokémon no jogo.
- Tipos: Os Pokémon têm um ou dois tipos, que determinam suas fraquezas e resistências a diferentes tipos de ataques. Existem 18 tipos diferentes no total, incluindo tipos como grama, fogo, água, elétrico, psíquico, dragão, entre outros. A combinação de tipos de um Pokémon pode afetar significativamente sua eficácia em batalha. Exemplo: Um Pokémon do tipo grama recebe dano dobrado de um ataque tipo fogo, mas recebe metade do dano de um pokémon tipo água. A combinação de dois tipos pode trazer imunidades diferentes e resistências de até 1/4 de dano, mas também receber 4 vezes o dano de determinados tipos.
- Características físicas: Altura e peso também podem influenciar nas batalhas, pois existem golpes e habilidades que levam em consideração essas características
- Porcentagem na natureza de machos e fêmeas: Alguns Pokémon têm diferenças de gênero em suas proporções de machos e fêmeas na natureza. Enquanto alguns Pokémon têm uma divisão de gênero de 50/50, outros podem ter uma proporção maior de machos ou fêmeas. Isso pode afetar a criação e a obtenção de Pokémon com determinados atributos ou movimentos, além da eficácia de movimentos que usam esse atributo com base.
- Pokémon sem gênero: Alguns Pokémon são considerados sem gênero e não têm distinção entre macho e fêmea. Isso geralmente se aplica a Pokémon lendários, míticos e alguns Pokémon semelhantes a objetos.

Essas características únicas contribuem para a diversidade e complexidade do mundo Pokémon, tornando cada espécie única e oferecendo uma ampla gama de estratégias e possibilidades no jogo.

Além disso existem características (também únicas a cada espécie ou até indivíduo) que dão mais informações sobre o contexto em que ele foi inserido

- Gerações: As gerações referem-se aos diferentes períodos de lançamento dos jogos principais da série Pokémon. Cada geração, geralmente, coincide com o lançamento de novos jogos principais da Nintendo. Cada geração introduz novos Pokémon, mecânicas de jogabilidade, regiões exploráveis e histórias. Atualmente a franquia possui 9 gerações.
- Lendário/Sub Lendário/Mítico: Esses termos são usados para descrever certos Pokémon que são considerados especiais ou únicos dentro do universo Pokémon. Os Pokémon lendários são poderosos e geralmente têm um papel significativo na narrativa do jogo. Os sub lendários são uma categoria menos restrita, mas ainda incluem Pokémon raros e poderosos, com participações em histórias secundárias. Os Pokémon míticos são extremamente raros e associados à lendas e mitos no universo, e geralmente só podem ser obtidos através de eventos especiais ou distribuição online.
- Catch Rate: A taxa de captura é uma medida da facilidade com que um jogador pode capturar um Pokémon selvagem durante uma batalha. Quanto mais alta a taxa de captura, mais fácil é capturar o Pokémon. A taxa de captura varia de acordo com o Pokémon e pode ser influenciada por fatores como o status de saúde do Pokémon selvagem, o uso de Poké Balls específicas e outras condições.
- Base Experience: A experiência base é a quantidade de experiência que um jogador recebe ao derrotar um Pokémon selvagem. Cada Pokémon tem uma quantidade específica de experiência base associada a ele, que é usada para calcular a quantidade de experiência que os Pokémon do jogador recebem após uma batalha.
- Egg Cycles: Os ciclos de ovos representam o tempo necessário para que um ovo de Pokémon choque e produza um Pokémon recém-nascido. Cada espécie de Pokémon tem um número específico de ciclos de ovos necessários para chocar, que pode variar de algumas centenas a milhares de ciclos, dependendo da espécie. Um ciclo possui 255-257 passos (dependendo do jogo)








# 2. Dicionário Dataset

O DataFrame consiste em uma lista de todos os pokémon existentes até a 8ª geração (final de 2020). No total, são 898 Pokémon distintos, contando com outras formas da mesma espécie.

O data frame pode ser baixado em: https://www.kaggle.com/datasets/mariotormo/complete-pokemon-dataset-updated-090420

O dataframe originalmente contém 51 colunas, que são:

- index: uma coluna de index
- pokedex_number: número do pokémon na National Pokédex (inteiro)
- name: nome em inglês do Pokémon (texto)
- german_name: nome em alemão do Pokémon (texto)
- japanese_name: nome original em japonês do Pokémon (texto)
- generation: o número da geração em que o pokémon foi originalmente descoberto (inteiro: de 1 a 8)
- status: diz se aquela espécie é um pokémon normal, sub lendário, lendário ou mítico (texto: Normal, Sub Legendary, Legendary, Mythical)
- species: a categoria (texto)
- type_number: números de tipo que o pokémon tem (inteiro: 1 ou 2)
- type_1: primeiro tipo do pokémon (texto)
- type_2: segundo tipo do pokémon, se ele o tiver (texto, ou vazio se não tiver)
- height_m: altura em metros (float)
- weight_kg: Tpeso em quilogramas (float)
- abilities_number: quantidade de habilidades que a espécie possui (inteiro: 1, 2 ou 3)
- ability_1: nome da primeira habilidade (texto)
- ability_2: nome da segunda habilidade, se ele tiver (texto)
- ability_hidden: nome da habilidade escondida, se ele tiver (texto)
- total_points: soma dos status base - HP, attack, defesa, special attack, special defense, speed (inteiro)
- hp: o HP base do pokémon (inteiro)
- attack: o ataque base do pokémon (inteiro)
- defense: a defesa base do pokémon (inteiro)
- sp_attack: o ataque especial base do pokémon (inteiro)
- sp_defense: a defesa especial base do pokémon (inteiro)
- speed: a velocidade base do pokémon (inteiro)
- catch_rate: o catch rate do pokémon (inteiro)
- base_friendship: a amizade base do pokémon quando é capturado (inteiro)
- base_experience: a experiência base do pokémon quando é capturado (inteiro)
- growth_rate: o tipo da curva de crescimento de experiência (texto)
- egg_type_number: números de grupos de tipo de ovo que o pokémon possui (inteiro: 0, 1 ou 2)
- egg_type_1: nome do primeiro egg type que o pokémon possui (texto)
- egg_type_2: nome do segundo egg type que o pokémon pode ter (texto)
- percentage_male: o percentual de indivíduos macho daquela espécie na natureza, vazio se o pokémon não tiver gênero (texto)
- egg_cycles: número de cilos (255-257 passos) necessários para rachar o ovo do pokémon (inteiro)
- against_X: dezoito características que apresentam um multiplicador de dano sofrido contra um ataque de um tipo específico (float)

# 3. Importando dataset

In [1]:
# importanto pandas
import pandas as pd
import numpy as np

In [2]:
# importando dataset
!git clone https://github.com/brunof-santos/cesar-school-imersao-python.git

Cloning into 'cesar-school-imersao-python'...
remote: Enumerating objects: 24, done.
remote: Counting objects: 100% (24/24), done.
remote: Compressing objects: 100% (21/21), done.
remote: Total 24 (delta 5), reused 0 (delta 0), pack-reused 0
Receiving objects: 100% (24/24), 79.86 KiB | 962.00 KiB/s, done.
Resolving deltas: 100% (5/5), done.


In [3]:
# lendo dataframe
df_pokemon = pd.read_csv('/content/cesar-school-imersao-python/2-introducao-a-analise-de-dados/pokedex_(Update_04.21).csv')

# 4. Tratamento de dados

In [4]:
df_pokemon.head()

Unnamed: 0  pokedex_number           name german_name        japanese_name  \
0           0               1      Bulbasaur     Bisasam  フシギダネ (Fushigidane)   
1           1               2        Ivysaur   Bisaknosp   フシギソウ (Fushigisou)   
2           2               3       Venusaur    Bisaflor  フシギバナ (Fushigibana)   
3           3               3  Mega Venusaur    Bisaflor  フシギバナ (Fushigibana)   
4           4               4     Charmander    Glumanda      ヒトカゲ (Hitokage)   

   generation  status         species  type_number type_1  ... against_ground  \
0           1  Normal    Seed Pokémon            2  Grass  ...            1.0   
1           1  Normal    Seed Pokémon            2  Grass  ...            1.0   
2           1  Normal    Seed Pokémon            2  Grass  ...            1.0   
3           1  Normal    Seed Pokémon            2  Grass  ...            1.0   
4           1  Normal  Lizard Pokémon            1   Fire  ...            2.0   

   against_flying  against_psychic  against_bug against_rock against_ghost  \
0             2.0              2.0          1.0          1.0           1.0   
1             2.0              2.0          1.0          1.0           1.0   
2             2.0              2.0          1.0          1.0           1.0   
3             2.0              2.0          1.0          1.0           1.0   
4             1.0              1.0          0.5          2.0           1.0   

  against_dragon  against_dark  against_steel  against_fairy  
0            1.0           1.0            1.0            0.5  
1            1.0           1.0            1.0            0.5  
2            1.0           1.0            1.0            0.5  
3            1.0           1.0            1.0            0.5  
4            1.0           1.0            0.5            0.5  

[5 rows x 51 columns]

Analisando o dataframe, definimos 4 perguntas a serem respondidas:

1 - Qual tipo de pokémon que mais aparece?

2 - Qual a habilidade que mais aparece?

3 - Qual é mediana poder de cada um dos tipos primarios dos pokemons?

4 - Quais são os 20 pokemons com maior ataque

Para o tratamento de dados, iremos:
- apagar a primeira coluna sem nome
- apagar a coluna german_name, pois é uma informação inútil
- apagar a coluna japanese_name, pois é uma informação inútil
- apagar as 18 últimas colunas against_X, pois não é uma informação útil no momento
- criar coluna percentage_female, que será 100-percentage_male, caso percentage_male não esteja vazia (obs: existe percentage_male 0, não confundir com vazio)
- criar coluna onde será definido se o pokemon possui gênero ou não

In [5]:
# primeiro verificamos as colunas exisentes no dataset
df_pokemon.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1045 entries, 0 to 1044
Data columns (total 51 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Unnamed: 0        1045 non-null   int64  
 1   pokedex_number    1045 non-null   int64  
 2   name              1045 non-null   object 
 3   german_name       1045 non-null   object 
 4   japanese_name     1045 non-null   object 
 5   generation        1045 non-null   int64  
 6   status            1045 non-null   object 
 7   species           1045 non-null   object 
 8   type_number       1045 non-null   int64  
 9   type_1            1045 non-null   object 
 10  type_2            553 non-null    object 
 11  height_m          1045 non-null   float64
 12  weight_kg         1044 non-null   float64
 13  abilities_number  1045 non-null   int64  
 14  ability_1         1042 non-null   object 
 15  ability_2         516 non-null    object 
 16  ability_hidden    813 non-null    object 


In [6]:
# apagando colunas
df_pokemon = df_pokemon.drop(['Unnamed: 0', 'german_name', 'japanese_name', 'against_normal', 'against_fire', 'against_water', 'against_electric','against_grass', 'against_ice', 'against_fight', 'against_poison', 'against_ground', 'against_flying', 'against_psychic', 'against_bug', 'against_rock', 'against_ghost', 'against_dragon', 'against_dark', 'against_steel', 'against_fairy'], axis=1)
# verificando as colunas que restaram e confirmando se as desejadas foram excluídas
df_pokemon.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1045 entries, 0 to 1044
Data columns (total 30 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   pokedex_number    1045 non-null   int64  
 1   name              1045 non-null   object 
 2   generation        1045 non-null   int64  
 3   status            1045 non-null   object 
 4   species           1045 non-null   object 
 5   type_number       1045 non-null   int64  
 6   type_1            1045 non-null   object 
 7   type_2            553 non-null    object 
 8   height_m          1045 non-null   float64
 9   weight_kg         1044 non-null   float64
 10  abilities_number  1045 non-null   int64  
 11  ability_1         1042 non-null   object 
 12  ability_2         516 non-null    object 
 13  ability_hidden    813 non-null    object 
 14  total_points      1045 non-null   int64  
 15  hp                1045 non-null   int64  
 16  attack            1045 non-null   int64  


In [7]:
# criando coluna de percentage female
df_pokemon['percentage_female'] = 100-df_pokemon['percentage_male']
# checando se o cálculo está correto e se valores NaN não buagaram algo (obs: sabemos que o Magnemite não possui gênero e o Charmander possui 87,5% de chance de ser macho)
df_pokemon[(df_pokemon['name'] == 'Magnemite') | (df_pokemon['name'] == 'Charmander')]

pokedex_number        name  generation  status         species  \
4                 4  Charmander           1  Normal  Lizard Pokémon   
108              81   Magnemite           1  Normal  Magnet Pokémon   

     type_number    type_1 type_2  height_m  weight_kg  ...  catch_rate  \
4              1      Fire    NaN       0.6        8.5  ...        45.0   
108            2  Electric  Steel       0.3        6.0  ...       190.0   

    base_friendship base_experience  growth_rate  egg_type_number  egg_type_1  \
4              70.0            62.0  Medium Slow                2      Dragon   
108            70.0            65.0  Medium Fast                1     Mineral   

     egg_type_2  percentage_male  egg_cycles  percentage_female  
4       Monster             87.5        20.0               12.5  
108         NaN              NaN        20.0                NaN  

[2 rows x 31 columns]

In [8]:
# adicionando coluna de genderless
df_pokemon['genderless'] = np.where(df_pokemon['percentage_male'].notna(), False, True)
# checando se a valoração está correta
df_pokemon[(df_pokemon['name'] == 'Magnemite') | (df_pokemon['name'] == 'Charmander')]

pokedex_number        name  generation  status         species  \
4                 4  Charmander           1  Normal  Lizard Pokémon   
108              81   Magnemite           1  Normal  Magnet Pokémon   

     type_number    type_1 type_2  height_m  weight_kg  ...  base_friendship  \
4              1      Fire    NaN       0.6        8.5  ...             70.0   
108            2  Electric  Steel       0.3        6.0  ...             70.0   

    base_experience  growth_rate egg_type_number  egg_type_1  egg_type_2  \
4              62.0  Medium Slow               2      Dragon     Monster   
108            65.0  Medium Fast               1     Mineral         NaN   

     percentage_male  egg_cycles  percentage_female  genderless  
4               87.5        20.0               12.5       False  
108              NaN        20.0                NaN        True  

[2 rows x 32 columns]

# 5. Visualização de dados

1 - Qual o tipo de pokémon que mais aparece?

Para responder essa pergunta não é possível apenas contar os tipos, pois existem duas colunas (type_1 e type_2). Para isso, foi necessário primeiro realizar a concatenação das colunas de tipos e em seguida realizar a contagem usando o value_counts

In [ ]:
# concatenando os tipos
tipos_concatenados = pd.concat([df_pokemon['type_1'], df_pokemon['type_2']])
# contando os tipos
contagem_tipos = tipos_concatenados.value_counts()
# mostrando a contagem de quantos pokémon existem para cada tipo
print(contagem_tipos)

Water       154
Normal      126
Psychic     121
Flying      120
Grass       119
Bug          90
Fire         82
Ground       80
Poison       79
Rock         75
Fighting     74
Dark         73
Electric     73
Dragon       73
Steel        71
Ghost        68
Fairy        63
Ice          57
Name: count, dtype: int64


Com isso conseguimos ver que o tipo que mais aparece é o tipo Water (Água) com 154 no total

2 - Qual a habilidade que mais aparece?

Para responder essa pergunta, também não é possível apenas contar as habilidades, pois existem três colunas (ability_1, ability_2 e	ability_hidden). Para isso, foi necessário primeiro realizar a concatenação das colunas e em seguida realizar a contagem usando o value_counts

In [ ]:
# juntando as colunas usando o melt
habilidades = df_pokemon.melt(value_vars=['ability_1', 'ability_2', 'ability_hidden'])['value']
# contando e juntando
contagem_habilidades = habilidades.value_counts()
# mostrando as 5 que mais aparecem
contagem_habilidades.head(5)

value
Swift Swim     43
Sturdy         41
Levitate       41
Keen Eye       40
Inner Focus    36
Name: count, dtype: int64

Após a análise, conseguimos encontrar que Swift Swim é a habilidade mais encontrada entre os pokémon, com o total de 43 aparições.

3 - Qual é mediana do poder de cada um dos tipos primários dos pokemons?

In [ ]:
# criano nova coluna com os tipos combinados e separados por _
df_pokemon['combined_types'] = df_pokemon['type_1'] + '_' + df_pokemon['type_2']

# se algum tipo for NaN, preencha com o valor da primeira coluna
df_pokemon['combined_types'].fillna(df_pokemon['type_1'], inplace=True)
# checando se de fato está criado e preenchido corretamente
print(df_pokemon.head(10))

In [30]:
# criando uma lista com todos os tipos
lista_tipos = ['Normal', 'Fighting', 'Flying', 'Poison', 'Ground', 'Rock', 'Bug', 'Ghost', 'Steel', 'Fire', 'Water', 'Grass', 'Electric', 'Psychic', 'Ice', 'Dragon', 'Dark', 'Fairy']
#criando um novo dataframe com mediana de cada tipo
mediana_tipos = pd.DataFrame(columns=['tipo', 'mediana'])
# vendo se de fato está criado
mediana_tipos.head(5)

Empty DataFrame
Columns: [tipo, mediana]
Index: []

In [31]:
# a ideia é procurar cada tipo na lista de tipos, filtrar o dataframe pra aquele
# tipo e fazer a mediana
for tipo in lista_tipos:
    # filtrando o dataframe onde o tipo está presente na coluna combined_types
    dataframe_filtrado = df_pokemon[df_pokemon['combined_types'].str.contains(tipo, na=False)]

    # calculando a mediana da coluna total_points para o tipo específico
    mediana = dataframe_filtrado['total_points'].median()
    # criando o conjunto que vai ser adicionado no dataframe de medianas
    conjunto = {'tipo': tipo, 'mediana': mediana}

    # adicionando o conjunto de tipo e mediana ao dataframe de medianas
    mediana_tipos = pd.concat([mediana_tipos, pd.DataFrame([conjunto])], ignore_index=True)

mediana_tipos = mediana_tipos.sort_values(by='mediana',ascending=False)
print(mediana_tipos)

        tipo  mediana
15    Dragon    540.0
8      Steel    495.0
1   Fighting    492.5
13   Psychic    490.0
5       Rock    485.0
9       Fire    482.0
14       Ice    480.0
7      Ghost    480.0
12  Electric    475.0
16      Dark    475.0
17     Fairy    470.0
2     Flying    470.0
10     Water    457.0
4     Ground    450.0
11     Grass    425.0
3     Poison    420.0
0     Normal    420.0
6        Bug    395.0


In [ ]:
# Criando um df que recebe um agrupamento dos tipos pela mediana dos total_points
media_tipos = df_pokemon.groupby('type_1')['total_points'].agg(mean=np.mean)

# Ordenando o DF das maiores medianas para as menores
media_tipos.sort_values(by='mean',ascending=False)

mean
type_1              
Dragon    536.365854
Psychic   489.370370
Steel     481.472222
Fire      455.015385
Electric  451.435484
Rock      448.683333
Fighting  446.404762
Ghost     444.071429
Flying    440.000000
Fairy     439.272727
Poison    435.512195
Dark      434.543478
Water     431.738806
Ground    431.146341
Ice       430.864865
Grass     418.241758
Normal    404.869565
Bug       384.617284

4 - Quais são os 20 pokemons com maior HP?

In [ ]:
# 20 Pokémon com os maiores valores na coluna hp
top_hp = df_pokemon.nlargest(20, 'hp')

# lista com todos os tipos
lista_tipos = ['Normal', 'Fighting', 'Flying', 'Poison', 'Ground', 'Rock', 'Bug', 'Ghost', 'Steel', 'Fire', 'Water', 'Grass', 'Electric', 'Psychic', 'Ice', 'Dragon', 'Dark', 'Fairy']

# criando um dicionário para contar a quantidade de cada tipo
contador_tipos= {type_: 0 for type_ in lista_tipos}

# contando a quantidade de cada tipo no dataframe
for type_ in lista_tipos:
    contador_tipos[type_] = (top_hp['type_1'] == type_).sum() + (top_hp['type_2'] == type_).sum()

# montando um df a partir do dicionário contador
contador_tipos_df = pd.DataFrame(list(contador_tipos.items()), columns=['Tipo', 'Quantidade'])
contador_tipos_df = contador_tipos_df.sort_values(by='Quantidade',ascending=False)

print(contador_tipos_df)


        Tipo  Quantidade
15    Dragon           7
0     Normal           6
7      Ghost           4
13   Psychic           3
3     Poison           2
10     Water           2
1   Fighting           1
16      Dark           1
17     Fairy           1
8      Steel           1
4     Ground           1
2     Flying           1
6        Bug           0
11     Grass           0
12  Electric           0
5       Rock           0
14       Ice           0
9       Fire           0


Com isso, encontramos que dentre os 20 pokémon com maior HP, 7 deles são do tipo Dragão, sendo esse o tipo que mais aparece.